### PR TITLE
[None][feat] GVR V2 decode top-k goes hint-free by default: the bracket comes from the current row

### PR DESCRIPTION
**Pull request by longcheng-nv**
## Summary

This PR makes GVR V2 decode Top-K **hint-free in production**. `run_varlen(logits, kv_lens, indices, ...)` derives the search bracket from the current logits row and `k` from `indices.shape[1]`; it no longer accepts, reads, updates, or seeds a previous-step Top-K tensor.

This removes prior-state storage and lifecycle requirements, and covers cases where a valid prior cannot exist (first decode step, disaggregated generation, or deployments without a per-layer prior buffer). Final validation shows no correctness or accuracy regression, while hint-free V2 reduces end-to-end TPOT by **6–19%** versus the production insertion/radix path across all six tested model/MTP cells.

### What changes

- **Current-row bracketing:** register-resident families fold a bracket from values already loaded from the current row; streaming families derive it from a deterministic sparse sample.
- **Simpler production API/state:** `run_varlen` drops `pre_idx` and `engine`; `TopK.needs_gvr_prior` remains true only for temporal implementations, so V2 skips prefill prior seeding.
- **One production mode:** `TRTLLM_GVR_V2_HINTED` and the hinted/hint-free launcher cache-key axis are removed; `hint_free` is the uniform kernel compile flag.
- **Exact fallback:** unsupported score layouts fall back to insertion/radix Top-K, not temporal CUDA GVR. Exhausted brackets, capacity overflow, and count-crossing failures use exact candidate or whole-row refinement.
- **Compatibility:** rows with `n <= k` keep identity indices plus `-1` padding. The compiled ABI retains its prior-tensor slot, but production aliases the output tensor into that slot and never reads it. Batch-uniform legacy entry points remain for tests and benchmarks only.

## Enablement and dispatch

```yaml
# --extra_llm_api_options
# algorithm: dsa for DeepSeek-V3.2
#            deepseek_v4 for DeepSeek-V4 Flash / Pro
sparse_attention_config:
    algorithm: deepseek_v4
    enable_heuristic_topk: true
```

| Path | Additional selection |
|---|---|
| GVR V2 hint-free (this PR) | `TRTLLM_GVR_SELF_SAMPLING=1` |
| GVR V1 temporal, CUDA | none |
| GVR V1 temporal, CuTe DSL | `use_cute_dsl_topk: true` |
| Production insertion/radix | `enable_heuristic_topk: false` |

V2 additionally requires CuTe DSL, SM100/103, checkpoint Top-K width in `{512, 1024, 2048}`, and compression ratio `1` or `4`. If the rollout variable is set but a V2 prerequisite is unmet, dispatch remains on V1. Once V2 is selected, a failed FP32/stride/alignment format gate falls back to exact insertion/radix. An active V2 logs `self-sampling GVR top-K engaged (... hint-free)` once.

### At a glance

| Area | Result |
|---|---|
| Operator correctness | **19,492 / 19,492** tie-aware exactness checks passed over 9,746 real-capture cases, batch sizes 1–1024 |
| End-to-end stability | **104 runs**, zero failures, zero arm-proof violations |
| End-to-end TPOT | Hint-free V2 beats production insertion/radix in every cell (**6–19% lower**) and is equal to or faster than temporal V1 |
| Accuracy | Flash and Pro GSM8K results are within one standard error across all three arms |
| CI | [`L0_MergeRequest_PR` #58098](https://github.com/NVIDIA/TensorRT-LLM/pull/18410#issuecomment-5508335720): **SUCCESS** on `4fa138bf` |

### End-to-end TPOT

Setup: 8×B200 hosts, TEP8, SWE-bench-64K entry #1, ISL 68,656, OSL 1,024, BS = concurrency = 1. `ori` is production insertion/radix, `v1` is temporal GVR, and `hf` is this PR. Speedup is baseline TPOT divided by candidate TPOT, so values above 1.0 mean the candidate is faster. Values are geomeans of per-repetition speedups. These are the [final results](https://github.com/NVIDIA/TensorRT-LLM/pull/18410#issuecomment-5512513101) and supersede the earlier interim comment.

| Model / MTP | hf speedup vs ori (ori/hf) | hf speedup vs v1 (v1/hf) | v1 speedup vs ori (ori/v1) | Pairs | TPOT ms (hf / v1 / ori) |
|---|---:|---:|---:|---:|---|
| DeepSeek-V3.2 MTP0 | **1.229×** (±0.2%) | 1.143× | 1.075× | 5 | 9.65 / 11.04 / 11.87 |
| DeepSeek-V3.2 MTP3 | **1.166×** (±2.4%) | 1.111× | 1.049× | 4 | 4.91 / 5.44 / 5.71 |
| DeepSeek-V4 Flash MTP0 | **1.112×** (±1.0%) | 1.024× | 1.087× | 7 | 5.58 / 5.71 / 6.20 |
| DeepSeek-V4 Flash MTP3 | **1.092×** (±1.0%) | 1.031× | 1.059× | 6 | 2.65 / 2.73 / 2.90 |
| DeepSeek-V4 Pro MTP0 | **1.101×** (±0.4%) | 1.019× | 1.081× | 6 | 9.57 / 9.76 / 10.55 |
| DeepSeek-V4 Pro MTP3 | **1.060×** (±1.5%) | 1.007× | 1.054× | 6 | 4.50 / 4.53 / 4.77 |

The V3.2 cells benefit most because every DSA layer runs the indexer and `K=2048` makes decode Top-K a larger share of the step. Results span the pre-merge `c1aa1eac` and post-merge `4fa138bf` heads (including #18501) with consistent per-cell ratios.

The CUDA Graph gauntlet also passed on Flash and Pro: hint-free MTP0/MTP3 at BS1, plus hint-free and V1 MTP3 at concurrency 8 with multi-bucket capture. Every run verified the engaged/mode log, no format-gate fallback, and no capture-time warmup miss.

### Operator performance

- Against the retired hinted mode on the full operator grid, `hf/hinted` latency is 0.992 geomean, 1.025 p90, and 1.15 max.
- In register families, where a valid hint had the largest potential effect, `hf/hinted` is 0.93–1.01; an invalid hint previously cost 3.6–6.0×.
- On three synthetic correlated-step shapes (`K=1024`, compression ratio 4, single B200), V2 has **2.16–2.26× lower average latency** than the legacy CuTe DSL temporal operator, with exact output on every shape.

### Accuracy

GSM8K strict match, TP=EP=8, MTP3, 1,319 samples, one run per arm:

| Arm | V4 Flash (%) | V4 Pro (%) |
|---|---:|---:|
| `ori` | 94.77 | 94.88 |
| `v1` | 95.22 | 95.15 |
| `hf` | 95.38 | 94.92 |

All arms are within one standard error (~0.6) on both models. Earlier V3.2 hint-free/hinted results also matched within run-to-run variation: GSM8K 96.51/96.75, MMLU 89.19/88.98, and GPQA-Diamond 72.90/71.04.

### Tests

- `test_gvr_selfsampling_topk.py`: **77 passed** — per-family varlen exactness against `torch.topk`, heterogeneous lengths, MTP row windows, short/zero-window rows, CUDA Graph capture/replay, dispatch, and cache keys.
- `test_top_k.py`: **13 passed**.
- DSA suite: **110 passed, 4 skipped**.
- Full CI on the current head: **SUCCESS**.

## Scope and follow-up

Engine selection remains unchanged; `TRTLLM_GVR_SELF_SAMPLING=1` is the rollout switch in this PR. Attention metadata may still allocate the prior arena whenever heuristic Top-K is enabled. Stacked follow-up #18446 replaces the environment switch with config-only two-level dispatch, removes the CUDA GVR heuristic, and stops allocating temporal prior state for V2.

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: There are two naming styles for this flag, `hinted` and `hint_free`, it seems that `hinted` is used in module level, `hint_free` is used in kernel level, can we unify them?

**Review by mingyangHao — APPROVED** [review]
Have tested this PR locally and passed.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py
Comment (longcheng-nv) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: Agreed. The next update standardizes this on `hint_free`. It also removes the production hinted mode and `TRTLLM_GVR_V2_HINTED`, so there is no longer a positive/negative flag split. The remaining temporal path is limited to test and benchmark coverage and is named explicitly.

Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: Whether the next update in this PR or in a follow-up refactor PR?

Comment (longcheng-nv) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: In this PR. I’ve folded it into the pending update: production V2 will be hint-free only, and the module/kernel naming will consistently use `hint_free`. I’ll push it together with the rebase after #18501 lands.

Comment (longcheng-nv) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: The follow-up PRs based on this PR for decoding top-K will include: 

- (1) unified the GVR V1(temporal hint) and GVR V2(self-sampling);
- (2) deprecate the old GVR code (CUDA GVR)

**Comment by longcheng-nv** [comment]
Pushed `c1aa1eac`: production V2 is now hint-free only, per the review discussion. `TRTLLM_GVR_V2_HINTED` and the hinted/hint-free mode axis are gone (the launcher cache key drops its mode bit), `run_varlen` loses its `pre_idx` and `engine` parameters, the kernel-side flag is renamed `hf` -> `hint_free`, the V2 hardware-format fall-through now lands on the exact insertion/radix op instead of the CUDA GVR path, and prefill prior seeding is skipped for V2 via `TopK.needs_gvr_prior`. On this head on B200: kernel suite `test_gvr_selfsampling_topk.py` 77 passed, module dispatch `test_top_k.py` 13 passed, DSA selection battery 13 passed. The rebase onto main follows once #18501 lands (it touches the same register family).

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py
Comment (longcheng-nv) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: Done in `c1aa1eac`: the module and kernel now use `hint_free` uniformly, and the production hinted mode itself is removed, so the flag only remains as the kernel's compile-time constant.

**Review by yuxianq — APPROVED** [review]

**Comment by longcheng-nv** [comment]
**Three-arm e2e results on this PR (3x 8xB200, TEP8, SWE-bench-64K entry#1 ISL 68,656, OSL 1024, BS=concurrency=1, 4-5 paired reps per cell).** Arms: `ori` = production radix, `v1` = temporal GVR, `hf` = this PR's hint-free V2. Run at `c1aa1eac28`; the subsequent merge commit only integrates main (incl. #18501) and does not touch the hint-free path.

TPOT geomean ratios (paired per rep; below 1.0 = faster):

| cell | hf/ori | hf/v1 | v1/ori | pairs |
|---|---:|---:|---:|---:|
| DSv4 Flash MTP0 | **0.893** (±0.7%) | 0.973 | 0.918 | 5 |
| DSv4 Flash MTP3 | **0.914** (±1.4%) | 0.965 | 0.947 | 4 |
| DSv4 Pro MTP0 | **0.910** (±0.4%) | 0.983 | 0.926 | 4 |
| DSv4 Pro MTP3 | **0.939** (±2.0%) | 0.987 | 0.951 | 4 |

Absolute TPOT means (ms): Flash MTP0 5.56/5.72/6.23 (hf/v1/ori), Flash MTP3 2.67/2.77/2.93, Pro MTP0 9.63/9.80/10.59, Pro MTP3 4.49/4.55/4.79.

- Hint-free V2 beats the production radix path by 6-11% TPOT on every cell and is consistently at or slightly ahead of the temporal V1 (hf/v1 0.965-0.987) — the prior-free default costs nothing.
- **Zero armproof violations across all 51 runs** (engaged/mode log strings, no format-gate fallback, no capture-time warmup misses) and zero failures.
- CUDA-graph gauntlet passed on both models: Flash hf MTP0/MTP3 BS1 + hf MTP3 conc8 multi-bucket capture + v1 MTP3 conc8; Pro hf MTP3 conc8.
- GSM8K (Flash, MTP3, 1319 samples, strict-match): ori 94.77 +/- 0.61, v1 95.22 +/- 0.59, **hf 95.38 +/- 0.58** — all three within 1SE, no accuracy regression.

Grid truncation notes: host reservations expired before reps 5-6 of some cells (4-5 pairs each still clear the >=4-pair gate); DSv3.2 perf e2e is blocked by an NVRTC `cuda.h` container issue unrelated to this PR (its exactness/perf is covered by the 9,746-case operator grid and the earlier V3.2 accuracy trio in the description).

**Review by zongfeijing — APPROVED** [review]

**Comment by longcheng-nv** [comment]
**Final three-arm e2e results (supersedes the interim table above).** Six cells, 104 runs on 8x B200 hosts, zero failures, zero arm-proof violations; per-rep paired TPOT ratios, geomean (lower is faster):

| Model / MTP | hf/ori | hf/v1 | v1/ori | pairs | TPOT ms (hf / v1 / ori) |
|---|---:|---:|---:|---:|---|
| DeepSeek-V3.2 MTP0 | **0.814** (±0.2%) | 0.875 | 0.930 | 5 | 9.65 / 11.04 / 11.87 |
| DeepSeek-V3.2 MTP3 | **0.858** (±2.4%) | 0.900 | 0.953 | 4 | 4.91 / 5.44 / 5.71 |
| DeepSeek-V4 Flash MTP0 | 0.899 (±1.0%) | 0.977 | 0.920 | 7 | 5.58 / 5.71 / 6.20 |
| DeepSeek-V4 Flash MTP3 | 0.916 (±1.0%) | 0.970 | 0.944 | 6 | 2.65 / 2.73 / 2.90 |
| DeepSeek-V4 Pro MTP0 | 0.908 (±0.4%) | 0.981 | 0.925 | 6 | 9.57 / 9.76 / 10.55 |
| DeepSeek-V4 Pro MTP3 | 0.943 (±1.5%) | 0.993 | 0.949 | 6 | 4.50 / 4.53 / 4.77 |

- Hint-free V2 beats the production insertion/radix path on every cell (6-19% TPOT) and never loses to the temporal V1; the V3.2 cells show the largest gain (every DSA layer runs the indexer and K=2048 makes decode Top-K a bigger share of the step).
- The earlier "DeepSeek-V3.2 NVRTC container blocker" turned out to be a false gate: per the op47 root-cause note, the trtllm-gen FMHA JIT locates its include layout through a worker-side `pip show tensorrt_llm`, which fails when the venv is not first on PATH — this campaign's runner already handles that, and the V3.2 engine runs cleanly in the same containers once the stale canary is bypassed.
- GSM8K parity (Pro added): ori 94.88 / v1 95.15 / hf 94.92 — within one standard error, same as Flash.
- Repetitions span the pre-merge (`c1aa1eac`) and post-merge (`4fa138bf`, incl. #18501) heads; per-cell ratios agree across the two, i.e. the main merge did not shift the envelope.

**Review by limin2021 — APPROVED** [review]

File: tests/unittest/_torch/thop/parallel/test_gvr_selfsampling_topk.py [hunk 1]
```diff
@@ -356,50 +373,9 @@ def _run_varlen_case(kv, next_n, cr, top_k, seed, with_values=False, engine="aut
...
-            ga = torch.sort(
-                torch.gather(logits[r], 0, out_a[r].to(torch.int64)) + 0.0, descending=True
-            ).values
-            gr = torch.sort(
-                torch.gather(logits[r], 0, out_r[r].to(torch.int64)) + 0.0, descending=True
-            ).values
-            assert torch.equal(ga, gr), f"row {r}: engine != reference"
+def test_selfsampling_topk_varlen(kv, next_n, cr, top_k):
+    """Validate the hint-free per-row KV-length and MTP window contract."""
+    _run_varlen_case(kv, next_n, cr, top_k, seed=sum(kv) + next_n + cr)
```
[the remark this answers is not in the document]
Comment (longcheng-nv) [reply tests/unittest/_torch/thop/parallel/test_gvr_selfsampling_topk.py:378]: Private test helper; deferring the annotation to avoid re-spinning CI on the merge head.
